### PR TITLE
[Performance Optimization] Add recompute for mHC fused_h_res_h_post_bda

### PR DESCRIPTION
**Pull request by DanielSun11**
### PR Category
Performance Optimization

### PR Types
Performance

### Description
把 `HyperConnectionModule.fused_h_res_h_post_bda` 包进 `RecomputeWithoutOutput` span，省掉 mHC 的 BDA kernel 在 backward 里固定的两块 fp32 上采样。

`FusedHPostBDA.save_for_backward` 在 `high_precision_mhc` 下会 pin 住 `[..., n, C]` 的 residual 上采样和 `[..., C]` 的 layer output 上采样，等于每个 half-layer 多两块 activation 量级的常驻显存。span 的 body 跑在 `no_grad` 下，内层 PyLayer 的 `save_for_backward` 不再保留任何东西，backward 时从低精度输入重放 kernel。

实现要点：

- `_fused_h_res_h_post_bda()`：只在能回本的地方建 span。`high_precision_mhc` 关闭时 kernel 保存的全是本来就活着的张量（reshape 是 view，没有上采样），此时建 span 只会白付一次重放，因此直接跳过。
- `h_res` / `h_post` 传入前 clone。这两个张量是 mHC aggregate span 的输出，会被 `Tensor._clear_data()` 清掉，而 BDA 的重放 hook 挂在下游、比 aggregate 的重放先触发，用 detach 出来的别名会读到空 holder。二者形状比 residual 小 C 倍，clone 的代价可忽略，也避免了依赖两个 span 的 hook 注册顺序。
- `preserve_rng_state` 只在 `dropout_prob > 0 and training` 时开：快路径无 dropout，重放本身是确定的，不必为 RNG 快照付钱。
- `_cast_and_discard_fused_bda()`：fp32 结果在 cast 回 layer dtype 之后就是死的（cast backward 只要 grad 不要 data），但它仍被 `span.outputs` 持有，不 discard 就等于用两块上采样换一块常驻 fp32 输出。discard 同时也是注册 `_recompute` 的唯一入口，所以是强制的而不是优化。dtype 本来就相同时 `Tensor.to()` 返回 self，此时先 clone 再 discard，避免清掉还在用的数据。

配套新增单卡测试 `tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py`（15 个 case）。span 只为省显存存在，所以要求它数值上完全免费：全部梯度与对应的无 span 基线用 `numpy.array_equal` 逐位比对，不用 `allclose`——容差比较会把真实回归（重放读到脏张量、RNG 状态丢失、上采样漏掉）当成噪声放过去。覆盖两种 BDA kernel（native / `fused_h_post_bda`）、有无 bias、`high_precision_mhc` 开关、fp32 / bf16、两层堆叠（用来复现 `h_res`/`h_post` 的 `_clear_data()` 别名问题）、整层走 `recompute_modules=['mhc_forward']`，以及 span 契约本身（该建时建、discard 强制、常驻显存确实下降）。

本地结果：`15 passed`。

### 是否引起精度变化
否。span 与基线逐位一致，已在上述测试中按 `numpy.array_equal` 断言。

### CODE REVIEW

File: tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+    rng = np.random.RandomState(seed)
+    resid_np = rng.randn(S, B, N * C).astype("float32") * 0.02
+    w_np = rng.randn(C, C).astype("float32") * 0.02
+    bias_np = rng.randn(C).astype("float32") * 0.02 if with_bias else None
+    return resid_np, w_np, bias_np
+
+
+def _use_fused_kernel(test_case, hc):
+    try:
+        from paddlefleet.fusions.fused_mhc_kernels import fused_h_post_bda
```
Comment (risemeup1111) [inline tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py]: P1

这里的导入不会在 cuTile 缺失时失败：fused_mhc_kernels.py 在该环境将 fused_h_post_bda 绑定为 _no_cutile_error，因此 ImportError 不会触发 skip；随后测试会在实际调用时抛出 RuntimeError。CUDA 但未安装 cuda.tile 的环境会因此让两个 fused 测试失败，而不是跳过。请在这里用 is_cutile_available() 判断并显式 skipTest，或对该能力做等价保护。

File: src/paddlefleet/transformer/transformer_layer.py [hunk 2]
```diff
@@ -1674,6 +1674,114 @@ def __init__(
...
+        bda_kwargs = {
+            "dropout_prob": self.hidden_dropout_prob,
+            "training": self.training,
+            "fused": self.config.bias_dropout_fusion,
+        }
+        # Without high_precision_mhc there is no up-cast: the kernel saves
+        # h_res/h_post/x and a *view* of original_residual (reshape does not
+        # copy), all of which are live anyway, so the span would buy nothing and
+        # only cost a replay.
+        if not self.config.high_precision_mhc:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py]: P1

这里仅用 high_precision_mhc 判断是否建立 span，但实际的 fp32 BDA fast path 还受 fused_h_res_h_post_bda() 的条件约束：关闭 FLAGS_use_accuracy_compatible_kernel，且训练时 dropout_prob == 0。当 FLAGS_use_accuracy_compatible_kernel=1 或训练开启 dropout 时，BDA 走 sequential 路径，没有这里要释放的 fp32 upcast；当前仍 clone h_res/h_post、重放整段 BDA，反而增加计算/内存开销。请让 enable 条件与该 fast-path 条件一致，并补测这两种配置。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [transformer_layer.py:1710](https://github.com/PaddlePaddle/PaddleFleet/pull/1742#discussion_r3776779188) | P1 | ✅ |
| 2 | [test_mhc_fused_h_res_h_post_bda_recompute.py:249](https://github.com/PaddlePaddle/PaddleFleet/pull/1742#discussion_r3776779187) | P1 | ✅ |

本轮按当前头 `f50b0f84d34d3db790d8bb2bcdd5fac55d276b8b` 复查了新增提交、两个 canonical threads、调用链和相关测试。低精度 active-dropout gate 已补齐并覆盖数值、gate 与显存回归；cuTile 能力保护也已修复，两项原 P1 均已解决。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+                layer, hc, "bfloat16", s=s, b=b, c=c
+            )
+            paddle.device.synchronize()
+            base = paddle.device.cuda.memory_allocated()
+            out, span = layer._fused_h_res_h_post_bda(
+                hc, h_res, resid, h_post, (x, None), wrap_bda
+            )
+            out = layer._cast_and_discard_fused_bda(out, resid.dtype, span)
+            paddle.device.synchronize()
+            used = paddle.device.cuda.memory_allocated() - base
```
[the remark this answers is not in the document]
Comment (DanielSun11) [reply tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py]: 已按建议改为 `paddle.device.memory_allocated()`（仓库内 `test_recompute_without_output.py`、`test_mla_gate_attn.py` 也是这个写法），ade5860f。

File: tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py (hunk 1 above)
Comment (DanielSun11) [reply tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py]: 确认成立，感谢指出。`fused_mhc_kernels.py` 在缺少 `cuda.tile` 时把 `fused_h_post_bda` 绑成了 `_no_cutile_error`，import 本身不会抛，所以原来的 `ImportError` 分支永远进不去，测试会在无 cuTile 环境下真的跑进错误分支而不是跳过。已改成用 `is_cutile_available()` 判断并 `skipTest`，ade5860f。

File: src/paddlefleet/transformer/transformer_layer.py (hunk 2 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/transformer_layer.py]: 这条**部分成立**，已经改了，但没有按“整条 sequential 路径都不包”来改，原因是实测 dropout 那一档是有收益的。

在 S=128, B=2, C=256, N=4, `high_precision_mhc=1` 下量了一遍 BDA 调用前后的 `paddle.device.memory_allocated()`（四档梯度都 bitwise 相同）：

| 配置 | 不包 span | 包 span | 节省 |
|---|---|---|---|
| bf16 fast path | 1808 KiB | 532 KiB | **+1276 KiB** |
| fp32 fast path | 1040 KiB | 1044 KiB | −4 KiB（打平） |
| fp32 dropout=0.1（sequential） | 1296 KiB | 1044 KiB | **+252 KiB** |
| fp32 accuracy-compatible kernel（sequential） | 1040 KiB | 1044 KiB | −4 KiB（打平） |

所以：

- `FLAGS_use_accuracy_compatible_kernel=1`：确实如你所说没有 fp32 up-cast 可省，纯粹多一次 replay，**已 veto**。
- dropout 生效的 sequential 路径：虽然没有 up-cast，但这条路径自己会存 `[..., n*C]` 中间量和 dropout mask，span 把这些藏起来了，实测省 252 KiB，如果一并 veto 会白丢这部分收益，所以**保留**（span 里带 `preserve_rng_state=True`，mask 在 replay 时能复现）。
- fp32 residual 时 fast path 的 up-cast 退化成 no-op，是“打平”而不是“变差”（−4 KiB 是 span 自身的记账开销），没有为它加 dtype 条件分支。

实现上把这个判断挪到了 kernel 旁边，作为 `HyperConnectionModule.bda_span_pays_off(dropout_prob, training)`，因为它依赖 `fused_h_res_h_post_bda` 走哪条路，放在 layer 里判断会重复 kernel 的内部条件。commit ade5860f。

单测同步补了这个 gate 的各个分支（`test_no_span_with_accuracy_compatible_kernel`、`test_accuracy_compatible_kernel_is_not_wrapped`、`test_span_created_when_dropout_is_active`、`test_span_created_in_eval_with_dropout_configured`）以及 sequential 路径的数值与显存两项（`test_sequential_path_with_dropout_*`、`test_span_shrinks_the_retained_set_on_the_sequential_path`）。

File: tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py (hunk 1 above)
Comment (risemeup1111) [reply tests/single_card_tests/transformer/test_mhc_fused_h_res_h_post_bda_recompute.py]: 已验证 `ade5860f`：测试现在先调用 `is_cutile_available()`，缺少 cuTile 时会显式 `skipTest`，不会再调用运行时 stub。该项已解决。

File: src/paddlefleet/transformer/transformer_layer.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py]: 已验证 `ade5860f` 对 accuracy-compatible 路径的 veto，以及 active dropout 下保留 span、`preserve_rng_state=True` 和对应显存/逐位数值测试；作者提到的两个 bf16 失败也属于不包 span 的既有路径问题，可另行跟踪，不阻塞本 PR。

但性能 gate 仍有一个遗漏：`src/paddlefleet/transformer/hyper_connection.py:674-678` 先无条件以 `high_precision_mhc=False` 返回 `False`，所以 `high_precision_mhc=False` 且训练中 `dropout_prob>0` 时，`_fused_h_res_h_post_bda()` 仍不会建立 span。此配置走同一个 sequential BDA 路径（`hyper_connection.py:737-750`），其 `[..., n*C]` 中间量和 dropout mask 仍可被 span 隐藏；同时 `forward()` 在 `high_precision_mhc=False` 时保留输入 dtype（`hyper_connection.py:544-549`），不会触发作者说明的 fp32/bf16 bmm mismatch。请把 active-dropout 判断置于 high-precision 早退之前，并为该低精度 dropout 配置补一条 gate/显存回归测试。该遗漏仍保留为 P1。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/transformer_layer.py (hunk 2 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/transformer_layer.py]: 确认成立，已修，commit f50b0f84。

`bda_span_pays_off` 里把 active-dropout 判断提到了两个早退之前：

```python
if dropout_prob > 0.0 and training:
    return True
if not self.config.high_precision_mhc:
    return False
return not _use_accuracy_compatible_kernel()
```

顺着这个思路又量了一遍（S=128, B=2, C=256, N=4, fp32, dropout=0.1，梯度全部 bitwise 相同）：

| 配置 | 不包 span | 包 span | 节省 |
|---|---|---|---|
| `high_precision_mhc=False` | 1296 KiB | 1044 KiB | **+252 KiB** |
| `high_precision_mhc=True` | 1296 KiB | 1044 KiB | +252 KiB |
| accuracy-compatible kernel ON | 1296 KiB | 1044 KiB | **+252 KiB** |

三档完全一样，说明这里省的其实就是 dropout mask：252 KiB ≈ `s*b*n*C` 字节 = 256 KiB 减去 span 自身 20 KiB 的 clone 开销，mask 是每个输出元素 1 字节，跟 mHC 精度和 accuracy-compatible 开关都无关。所以第三档（accuracy-compatible + dropout）之前也被误 veto 了，这次一并放开，注释和 docstring 也改成按 mask 来解释（之前写成“`[..., n*C]` 中间量”不准确，`mixed`/`x_expanded` 实际会被释放）。

补的测试：
- gate：`test_span_created_for_low_precision_dropout`、`test_span_created_for_accuracy_compatible_kernel_with_dropout`
- 数值：`test_low_precision_mhc_with_dropout_is_wrapped`（逐位）
- 显存：`test_span_shrinks_the_retained_set_without_high_precision_mhc`（与 sequential 那条共用 `_assert_sequential_span_saves`）

`test_low_precision_mhc_is_not_wrapped`（dropout=0）保持不变，仍应 veto。26 个测试全过。

有一处需要更正你的判断：**`high_precision_mhc=False` 并不能避开 bf16 的 bmm mismatch**。`forward()` 确实不再 up-cast `hidden_states`（返回的 `aggregated` 是 bf16），但 `h_res`/`h_post` 走的是 mapping+sinkhorn，输出恒为 fp32：

```
high_precision_mhc=True : agg=paddle.float32 h_res=paddle.float32 h_post=paddle.float32
high_precision_mhc=False: agg=paddle.bfloat16 h_res=paddle.float32 h_post=paddle.float32
```

所以 `apply_h_res` 里还是 `bmm(fp32 h_res, bf16 residual)`，实测在 `high_precision_mhc=False` + bf16 + dropout 下同样报 `ValueError ... (bfloat16) does not match ... (float32)`（`hyper_connection.py:468`），且不包 span 也一样报。因此新补的这些 sequential 相关测试仍然只能用 fp32，测试里写了注释说明原因。这个存量问题按你说的另行跟踪。

Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py]: 已验证 `f50b0f84`：active dropout 现在在两个 precision/accuracy gate 之前返回可建立 span，并新增了低精度 dropout 的逐位数值、gate 与显存回归测试；原性能 P1 已解决。

同时确认你补充的 dtype 说明：`high_precision_mhc=False` 时 `h_res` 仍为 fp32，bf16 dropout 的 `bmm` mismatch 是不包 span 也会触发的既有问题，不阻塞本 PR。

**Review by risemeup1111 — APPROVED** [review]

**Review by LiYuRio — APPROVED** [review]
